### PR TITLE
Add blog post sorting options for visitors

### LINKED ISSUE #196: Visitors cannot choose other post sorting options
## Problem

PR #186 makes the front end display blog posts in descending order of creation date, with no way for visitors to change the order. Visitors might want the oldest or most recently edited posts listed first.

## Suggestion

- Allow users to choose among several preset sorting options (e.g., default, by creation date, by title), and sort posts accordingly on the front end.
- Send no information to the back end.

### PR DESCRIPTION
**Pull request by flybot-nam-nguyenhoai**
## Closes #196

- [x] Allow visitors to sort blog posts by date created, date updated or title.
    The default is still sorting by date created, descending (newest first).

### CODE REVIEW

File: client/web/src/flybot/client/web/core/dom/page/options.cljs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+(ns flybot.client.web.core.dom.page.options
+  (:require [re-frame.core :as rf]))
+
+(defn blog-sorting-form
+  []
+  [:form
+   (into
+    [:fieldset]
+    [[:label {:for "sort-by"} "Sort posts by:"]
+     [:select#blog-sorting-options
```
Comment (skydread1) [inline client/web/src/flybot/client/web/core/dom/page/options.cljs:10]: - no need `into`.
- `fieldset ` needs a `legend`

File: client/web/src/flybot/client/web/core/dom/page/options.cljs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      {:name "Sort by"
+       :on-change #(rf/dispatch [:evt.page.form/set-blog-sorting-options
+                                 (.. % -target -value)])}
+      [:option
+       {:selected true
```
Comment (skydread1) [inline client/web/src/flybot/client/web/core/dom/page/options.cljs:15]: browser was raising a warning and using:
```clojure
:defaultValue true
```
should fix it.

**Review by skydread1 — CHANGES_REQUESTED** [review]

File: client/web/src/flybot/client/web/core/dom/page.cljs [hunk 3]
```diff
@@ -30,16 +31,30 @@
...
+                                        posts))
                        (sort-by :post/default-order posts))
         new-post {:post/id "new-post-temp-id"}]
     [:section.container
      {:class (name page-name)
       :key   (name page-name)}
      [:h1 page-name]
+     (when (= :blog page-name)
+       [:div.post [page.options/blog-sorting-form]])
+     [page-post new-post]
```
[the remark this answers is not in the document]
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/dom/page.cljs:55]: PR #200 removed `[:h1 "New Post"]` from the temporary new post, so now the new post shouldn't have any headings. `[page-post new-post]` just results in a big plus button in edit mode, or nothing at all in read mode.

this_pr_blog

If I use `blog-post-short` on `new-post` (like in #200), then there's a blank div between the page name and blog entries.

pr_200_blog

[the remark this answers is not in the document]
Comment (skydread1) [reply client/web/src/flybot/client/web/core/dom/page.cljs:55]: 1. click the `+` to create a post
2. write a post
3. preview your post
4. observe the broken css

File: client/web/src/flybot/client/web/core/dom/page/options.cljs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  (:require [re-frame.core :as rf]))
+
+(defn blog-sorting-form
+  []
+  [:form
+   (into
+    [:fieldset]
+    [[:label {:for "sort-by"} "Sort posts by:"]
+     [:select#blog-sorting-options
+      {:name "Sort by"
```
Comment (skydread1) [inline client/web/src/flybot/client/web/core/dom/page/options.cljs:11]: name should be `sort-by`

**Review by skydread1 — CHANGES_REQUESTED** [review]

File: client/web/src/flybot/client/web/core/dom/page.cljs (hunk 3 above)
[the remark this answers is not in the document]
Comment (skydread1) [reply client/web/src/flybot/client/web/core/dom/page.cljs:55]: And for the gap in the UI, just add a temp id check:

```clojure
(defn blog-post-short
  [{:post/keys [css-class id] :as post}]
  (let [post-title (web.utils/post->title post)]
    (when-not (utils/temporary-id? id)
      [:div.post.short
       {:key id
        :id id}
       [post-link post
        [:div.post-body
         {:class css-class}
         [:h2 post-title]
         [post-authors post]]]])))
````

[the remark this answers is not in the document]
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/dom/page.cljs:55]: With the code above, the new post button disappears in both edit mode and read mode. Fixing.

File: client/web/src/flybot/client/web/core/dom/page/options.cljs (hunk 2 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/dom/page/options.cljs:15]: Fixed.

Observations:

- Using either `value` or `defaultValue` works.
- `defaultValue` seems to be specific to React.
- The difference is that `value` is for "controlled elements" and `defaultValue` is for "uncontrolled elements."

File: client/web/src/flybot/client/web/core/dom/page/options.cljs (hunk 4 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/dom/page/options.cljs:11]: Fixed; name is now `blog-sorting-options`.

File: client/web/src/flybot/client/web/core/dom/page/options.cljs (hunk 1 above)
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/dom/page/options.cljs]: Fixed.

`(into)` was originally used with `(interpose [:br])` to insert breaks between fieldset elements. Both are now removed.

File: client/web/src/flybot/client/web/core/dom/page.cljs [hunk 5]
```diff
@@ -30,16 +31,30 @@
...
+                                                   {:app/blog-sorting '?x}])]
+                               (sort-by (case by
+                                          :date-created :post/creation-date
+                                          :date-updated :post/last-edit-date
+                                          :title web.utils/post->title
+                                          :post/creation-date)
+                                        (case direction
+                                          :ascending compare
+                                          #(compare %2 %1))
+                                        posts))
```
[the remark this answers is not in the document]
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/dom/page.cljs]: Fixed. I guess I really like `case`.

File: client/web/src/flybot/client/web/core/dom/page.cljs (hunk 3 above)
[the remark this answers is not in the document]
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/dom/page.cljs]: Restored according to PR #200.

File: client/web/src/flybot/client/web/core/db/event.cljs [hunk 6]
```diff
@@ -141,3 +142,11 @@
...
  (fn [db [_ new-match]]
    (assoc db :app/current-view new-match)))
+
+;; ------ Post sorting ------
+
+(rf/reg-event-db
+ :evt.page.form/set-blog-sorting-options
+ [(rf/path [:app/blog-sorting])]
+ (fn [_ [_ new-options]]
+   (edn/read-string new-options)))
```
[the remark this answers is not in the document]
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/db/event.cljs:152]: Should the form itself be in a new namespace called `flybot.client.common.dom.page.options`?

[the remark this answers is not in the document]
Comment (skydread1) [reply client/web/src/flybot/client/web/core/db/event.cljs]: No, the re-frame logic is the same for both platform (as mush as we can).

However, the markup differs between platforms, react native using native components.

[the remark this answers is not in the document]
Comment (flybot-nam-nguyenhoai) [reply client/web/src/flybot/client/web/core/db/event.cljs]: Ah, I misunderstood. So for this PR only the re-frame dispatch event should be moved to `common`.

I'll have to force-push, apologies.

**Review by skydread1 — APPROVED** [review]

**Review by skydread1 — APPROVED** [review]

**Pull request merged**